### PR TITLE
add compound-v2

### PR DESCRIPTION
**Pull request by dmelotik**
- [x] Add in schema 1.1.0
- [x] Accrue interest on inputBalances / totalBorrowNative
- [x] fix inputTokenBalances - (ie, market liquidity = (deposits - withdraws + interest))
- [x] fix outputTokenBalances - decrease on withdraw
- [x] figure out how to setCollateralActive
- [x] Find calculation error for totalBorrowUSD/totalDepositUSD
- [x] ensure all is working properly

### CODE REVIEW

**Review by Bastek8 — APPROVED** [review]

**Review by Bastek8 — APPROVED** [review]
.gitignore

File: subgraphs/compound-v2/src/mappings/helpers.ts [hunk 1]
```diff
@@ -0,0 +1,433 @@
...
+}
+
+/////////////////////////
+//// Updates Helpers ////
+/////////////////////////
+
+export function updateMarketPrices(market: Market, event: ethereum.Event): void {
+  let underlyingDecimals = getOrCreateToken(market.inputTokens[0]).decimals;
+  let mantissaFactorBD = exponentToBigDecimal(DEFAULT_DECIMALS);
+  let exchangeRate = getExchangeRate(Address.fromString(market.id), event);
```
[the remark this answers is not in the document]
Comment (dmelotik) [reply subgraphs/compound-v2/src/mappings/helpers.ts:339]: it's needed for exchangeRate and token price function calls

File: subgraphs/compound-v2/src/common/metrics.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+  let protocolSideRevenue = BIGDECIMAL_ZERO;
+  let feesUSD = BIGDECIMAL_ZERO; // aka Total revenue = market outstanding borrows * market borrow rate
+  for (let i = 0; i < protocol._marketIds.length; i++) {
+    let market = getOrCreateMarket(event, Address.fromString(protocol._marketIds[i]));
+    let underlyingDecimals = getOrCreateToken(market.inputTokens[0]).decimals;
+    let outstandingBorrowUSD = market._outstandingBorrowAmount
+      .toBigDecimal()
+      .div(exponentToBigDecimal(underlyingDecimals))
+      .times(market._inputTokenPrice);
+    supplySideRevenue = outstandingBorrowUSD
```
[the remark this answers is not in the document]
Comment (dmelotik) [reply subgraphs/compound-v2/src/common/metrics.ts]: i agree - fixed now with helper method in helpers.ts

File: subgraphs/compound-v2/src/common/metrics.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+      .times(market.variableBorrowRate)
+      .times(BIGDECIMAL_ONE.minus(market._reserveFactor))
+      .plus(supplySideRevenue);
+    protocolSideRevenue = outstandingBorrowUSD
```
[the remark this answers is not in the document]
Comment (dmelotik) [reply subgraphs/compound-v2/src/common/metrics.ts]: fixed - see above comment

File: subgraphs/compound-v2/src/common/utils/utils.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+export function exponentToBigDecimal(decimals: i32): BigDecimal {
+  let bigDecimal = BigDecimal.fromString("1");
+  for (let i = 0; i < decimals; i++) {
+    bigDecimal = bigDecimal.times(BigDecimal.fromString("10"));
+  }
+  return bigDecimal;
+}
+
+// get the amount in underlying token from cToken
+export function getExchangeRate(marketAddress: Address, event: ethereum.Event): BigInt {
```
[the remark this answers is not in the document]
Comment (dmelotik) [reply subgraphs/compound-v2/src/common/utils/utils.ts:15]: fixed

File: subgraphs/compound-v2/subgraph.yaml [hunk 5]
```diff
@@ -0,0 +1,85 @@
+specVersion: 0.0.2
+description: Compound is an open-source protocol for algorithmic, efficient Money Markets on the Ethereum blockchain.
```
[the remark this answers is not in the document]
Comment (dmelotik) [reply subgraphs/compound-v2/subgraph.yaml:2]: fixed

File: subgraphs/compound-v2/src/common/getters.ts [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+    let inputBalances = new Array<BigInt>();
+    inputBalances.push(BIGINT_ZERO);
+    marketMetrics.inputTokenBalances = inputBalances;
+    let inputPrices = new Array<BigDecimal>();
+    marketMetrics.inputTokenPricesUSD = inputPrices;
+    marketMetrics.outputTokenSupply = BIGINT_ZERO;
+    marketMetrics.outputTokenPriceUSD = BIGDECIMAL_ZERO;
+    let emissionsAmount = new Array<BigInt>();
+    emissionsAmount.push(BIGINT_ZERO);
+    marketMetrics.rewardTokenEmissionsAmount = emissionsAmount;
```
[the remark this answers is not in the document]
Comment (dmelotik) [reply subgraphs/compound-v2/src/common/getters.ts:86]: good catch - fixed

File: subgraphs/compound-v2/src/mappings/mappings.ts [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+  // get reserve factor
+  market._reserveFactor = event.params.newReserveFactorMantissa
+    .toBigDecimal()
+    .div(exponentToBigDecimal(DEFAULT_DECIMALS));
+  market.save();
+}
+
+export function handleNewCollateralFactor(event: NewCollateralFactor): void {
+  let market = getOrCreateMarket(event, event.params.cToken);
+  let newLTV = event.params.newCollateralFactorMantissa.toBigDecimal().div(exponentToBigDecimal(16));
```
Comment (this-username-is-taken) [inline subgraphs/compound-v2/src/mappings/mappings.ts:97]: can you make `16` a named constant and document it?

Comment (dmelotik) [reply subgraphs/compound-v2/src/mappings/mappings.ts]: fixed

Comment (this-username-is-taken) [reply subgraphs/compound-v2/src/mappings/mappings.ts]: this is fixed?

**Comment by this-username-is-taken** [comment]
Remove TODO?

**Review by this-username-is-taken — DISMISSED** [review]
Looks great Dylan! Just two minor changes and we are good to go 🚀 🚀 🚀

Feel free to merge it in yourself when it's ready

**Pull request merged**